Content.Server/Atmos/EntitySystems/GenericGasReactionSystem.cs

```diff
@@ -0,0 +1,115 @@
+using Content.Server.Atmos.Reactions;
+using Content.Shared.Atmos.Reactions;
+using Content.Shared.Atmos;
+using JetBrains.Annotations;
+using System.Collections;
+using System.Linq;
+
+namespace Content.Server.Atmos.EntitySystems;
+
+public sealed class GenericGasReactionSystem : EntitySystem
+{
+    [Dependency] private readonly AtmosphereSystem _atmosphere = default!;
+
+    /// <summary>
+    ///     Return a reaction rate (in units reactants per second) for a given reaction. Based on the
+    ///     Arrhenius equation (https://en.wikipedia.org/wiki/Arrhenius_equation).
+    ///
+    ///     This means that most reactions scale exponentially above the MinimumTemperatureRequirement.
+    /// </summary>
+    private float ReactionRate(GasReactionPrototype reaction, GasMixture mix, float dE)
+    {
+        float temp = mix.Temperature;
+
+        // Gas reactions have a MinimumEnergyRequirement which is in spirit activiation energy (Ea),
+        // but no reactions define it. So we have to calculate one to use. One way is to assume that
+        // Ea = 10*R*MinimumTemperatureRequirement such that Ea >> RT.
+        float TScaleFactor = 10;
+        float Ea = TScaleFactor*Atmospherics.R*reaction.MinimumTemperatureRequirement + dE;
+
+        // To compute initial rate coefficient A, assume that at temp = min temp we return 1/10.
+        float RateScaleFactor = 10; // not necessarily the same as TScaleFactor! Don't get confused!
+        float A = MathF.Exp(TScaleFactor) / RateScaleFactor;
+
+        // Prevent divide by zero
+        if (temp < Atmospherics.TCMB)
+            return 0;
+
+        return A*MathF.Exp(-Ea/(Atmospherics.R*temp));
+    }
+
+    /// <summary>
+    ///     Run all of the reactions given on the given gas mixture located in the given container.
+    /// </summary>
+    public ReactionResult ReactAll(IEnumerable<GasReactionPrototype> reactions, GasMixture mix, IGasMixtureHolder? holder)
+    {
+        float nTotal = mix.TotalMoles;
+        // Guard against very small amounts of gas in mixture
+        if (nTotal < Atmospherics.GasMinMoles)
+            return ReactionResult.NoReaction;
+
+        foreach (var reaction in reactions)
+        {
+            // Check if this is a generic YAML reaction (has reactants)
+            if (reaction.Reactants.Count == 0)
+                continue;
+
+            // Add concentration-dependent reaction rate
+            // For 1A + 2B -> 3C, the concentration-dependence is [A]^1 * [B]^2
+            float rate = 1f; // rate of this reaction
+            foreach (var (reactant, num) in reaction.Reactants)
+            {
+                float concentration = mix.GetMoles(reactant)/nTotal;
```

Comment:
```suggestion
                float concentration = mix.GetMoles(reactant)/mix.Volume;
```
The concentration should be in moles/volume, not the mole fraction. If you have the same gas amount in a bigger container, it should react slower since the molecules are farther apart and less likely to hit each other. The rates might need rebalancing for this.

Reply:
Possible divide by near-zero, perhaps multiplying by pressure would be better.